### DISCUSSION TITLE
QuickFIX/J Distribution - Maven Plugin Failure - `maven-assembly-plugin`

### DISCUSSION THREAD

**Discussion by vinipx** [quickfix-j/quickfixj#566] [Q&A] [answered]
[Not QuickFix/J specific, but might be happening to others]

I'm having failure in QuickFIX/J Distribution when trying to build project after cloning it:

[INFO] QuickFIX/J Distribution ............................ **FAILURE** [  4.636 s]

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:3.4.2:single (bin-assembly) on project quickfixj-distribution: Execution bin-assembly of goal org.apache.maven.plugins:maven-assembly-plugin:3.4.2:single failed.: NullPointerException -> [Help 1]
...
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:3.4.2:single (bin-assembly) on project quickfixj-distribution: Execution bin-assembly of goal org.apache.maven.plugins:maven-assembly-plugin:3.4.2:single failed.
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution bin-assembly of goal org.apache.maven.plugins:maven-assembly-plugin:3.4.2:single failed
...
Caused by: java.lang.NullPointerException
```

Maven and Java versions:
```
Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)
Maven home: C:\dev\apache-maven-3.8.6
Java version: 11.0.16.1, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-11.0.16.1

java --version
java 11.0.16.1 2022-08-18 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.16.1+1-LTS-1)
```

I've tried to use `mvn clean install` and even removing `.m2/repository`

Is anyone facing similar issue?

**Comment by vinipx** [comment]
I'm afraid it is connected to this plugin definition in pom.xml:
https://github.com/quickfix-j/quickfixj/blob/master/quickfixj-distribution/pom.xml#L149

**Comment by chrjohn** [comment, marked as answer]
Omit the`-PskipBundlePlugin` option.

**Reply by vinipx** [reply]
Thanks!

**Reply by chrjohn** [reply]
Don't know what triggers this, though. It was working before I did some updates to several Maven plugins, but don't know which exactly.

**Reply by vinipx** [reply]
Yeah, I couldn't find the root cause too. I'd suggest for us to handle in a separate bug issue maybe. 
I decided to publish here in discussion as someone else might face similar issue and can find workaround/fix meanwhile.

**Reply by chrjohn** [reply]
There was a similar problem here which manifests in a slightly different error message:
https://stackoverflow.com/a/73664485/4962355

**Comment by the-thing** [comment]
I also encoutered this problem while looking at something else.

The problem is `slf4j-jdk14` test scope dependency in quickfixj-core. There is a special case for it in distribution.xml assembly file. When running `skipBundlePlugin` profile this file doesn't exist.

```
[DEBUG] Adding dependency artifact org.slf4j:slf4j-jdk14:jar:2.0.6.
[DEBUG] Adding artifact: org.slf4j:slf4j-jdk14:jar:2.0.6 with file: null to assembly location: lib/slf4j-jdk14-2.0.6.jar.
```

I think the fix for this is to actually modify assembly instruction to exclude `slf4j-jdk14` from the build. I assume this is not required to be a part of `org.quickfixj-3.0.0-SNAPSHOT-bin.zip`. Is that correct? I'm not sure how this affects examples.

BTW. Sligtly different question. What are the expected differences in the generated artifacts when running with/without `skipBundlePlugin` profile?

**Reply by chrjohn** [reply]
Hi Marcin, welcome back. :)
Good spot, how did you find that out? I am no Maven expert and did not manage to look further into this problem. Looks like I needed to enable DEBUG information to find the problem?

I am not even sure we need the whole `slf4j-jdk14` logging dependency, which is a bridge to the `java.util.Logging` framework. As you said it is present in the tests, but we could probably also change that to use SLF4J directly.

----

`skipBundlePlugin` should prevent the execution of the `maven-bundle-plugin` which generates OSGi bundle information in the MANIFEST files. Since it takes some time (don't know exactly how long, it's years since I checked this) I introduced that profile to make the build faster.

**Reply by the-thing** [reply]
Yes. Adding -X parameter to mvn command will print the exception and the stack trace. I assume this dependency was/is required at some point to enable logging when running tests.

The fix is either to remove this dependency completely or just change distribution.xml file to exclude it from the quickfixj-distribution build. Both of these will result in org.slf4j:slf4j-jdk14:jar:2.0.6 missing org.quickfixj-3.0.0-SNAPSHOT-bin.zip lib directory.

**Reply by david-gibbs-ig** [reply]
@the-thing @chrjohn  Or add the dependency to the distribution module. I found that this works but running it locally. I thought it might be justifiable to add it to the distribution as it is a runtime dependency of the examples. 
https://github.com/quickfix-j/quickfixj/pull/460/commits/2af213f839452d4053ca65019efaeaec70901f55

**Reply by chrjohn** [reply]
@david-gibbs-ig I am actually not sure if `examples` actually requires `slf4j-jdk14`. It is listed in the `pom.xml` but did not find references to `java.util.logging` in the code of the `examples` module (unless `grep` did something wrong ;-)).

**Reply by chrjohn** [reply]
I just removed the `slf4j-jdk14` dependency from the examples and the build succeeded. I also was able to start the examples standalone-JARs (which seemed to be the reason why that dependency was included). There was a message that no default logging implementation was found and it was logged to console, which I think is OK.

**Reply by david-gibbs-ig** [reply]
@chrjohn , great, when you make a PR for that I will copy...

**Reply by chrjohn** [reply]
Hi @david-gibbs-ig 
here is the PR, if it succeeds I will merge to master.
https://github.com/quickfix-j/quickfixj/pull/615

**Reply by chrjohn** [reply]
@david-gibbs-ig merged #615

**Reply by david-gibbs-ig** [reply]
@chrjohn  thank you

**Comment by david-gibbs-ig** [comment]
I was just struggling with the same issue @the-thing , thanks for commenting. It was making me despondent. 

@chrjohn  I don't think I fully understand the bundle plugin but I think it's not helpful for the general (non OSGI) case. Can we consider removing it in version 3 ? We could perhaps make an example for users that want to use OSGI. 

I have made an experimental branch where I remove it from all but quickfixj-all but like I say I'm just experimenting. I think it need not be the responsibility of the QFJ project to build the bundles. If I understand what I have read about Apache Felix, an OSGI user could do this in their application project or in another project derived from QuickFIX/J. I think a good argument for what I am proposing is "How do we even know if QuickFIX/J is building the bundle correctly"?

**Reply by chrjohn** [reply]
Hmm, actually the `maven-bundle-plugin` should cause no problems. Maybe we could change it that is has to be explicitly enabled?
I don't know how much users really use this in an OSGi environment actually, but I know that there are some frameworks or app servers that use OSGi bundles (e.g. Apache Karaf, Glassfish/Payara).

> "How do we even know if QuickFIX/J is building the bundle correctly"

I am just relying on that the plugin works. Just as I am doing with the Javadoc plugin for example. :)

**Reply by david-gibbs-ig** [reply]
Hi, @chrjohn I think that the `maven-bundle-plugin` interferes with the normal packaging of the jars, even when that profile is not being run. This is why those duplicate class warnings occur. I've got a branch to see if I can show this is the case (working on the javadoc first). I think that  `maven-bundle-plugin` is an overhead for the QuickFIX/J project that isn't a library project's responsibility. I'm certainly no expert but QFJ 3 is a great opportunity for housekeeping and the way I see it, the bundle plugin imposes complexity on all users.

**Reply by chrjohn** [reply]
I don't know every library but in some of our projects we used OSGi bundles and were happy when we didn't need to transform the used JAR files to bundles first. It is no rocket science but also takes some time to get started and you have to repackage the JAR file and push it to some internal repository then.
From what I recall it could be that Apache libs were mostly OSGi bundles (probably because of Apache ServiceMix or Camel).
